packages/material-ui/src/InputBase/InputBase.js

```diff
@@ -142,291 +140,278 @@ export const styles = theme => {
  * It aims to be a simple building block for creating an input.
  * It contains a load of style reset and some state logic.
  */
-class InputBase extends React.Component {
-  static getDerivedStateFromProps(props, state) {
-    // The blur won't fire when the disabled state is set on a focused input.
-    // We need to book keep the focused state manually.
-    if (props.disabled && state.focused) {
-      return { focused: false };
+const InputBase = React.forwardRef(function InputBase(props, ref) {
+  const {
+    'aria-describedby': ariaDescribedby,
+    autoComplete,
+    autoFocus,
+    classes,
+    className: classNameProp,
+    defaultValue,
+    disabled,
+    endAdornment,
+    error,
+    fullWidth = false,
+    id,
+    inputComponent = 'input',
+    inputProps: { className: inputPropsClassName, ...inputPropsProp } = {},
+    inputRef: inputRefProp,
+    margin,
+    multiline = false,
+    name,
+    onBlur,
+    onChange,
+    onClick,
+    onEmpty,
+    onFilled,
+    onFocus,
+    onKeyDown,
+    onKeyUp,
+    placeholder,
+    readOnly,
+    renderPrefix,
+    rows,
+    rowsMax,
+    startAdornment,
+    type = 'text',
+    value,
+    ...other
+  } = props;
+
+  const { current: isControlled } = React.useRef(value != null);
+
+  const inputRef = React.useRef();
+  const handleInputRefWarning = React.useCallback(instance => {
+    warning(
+      !instance || instance instanceof HTMLInputElement || instance.focus,
+      [
+        'Material-UI: you have provided a `inputComponent` to the input component',
+        'that does not correctly handle the `inputRef` property.',
+        'Make sure the `inputRef` property is called with a HTMLInputElement.',
+      ].join('\n'),
+    );
+  }, []);
+  const handleInputPropsRefProp = useForkRef(inputPropsProp.ref, handleInputRefWarning);
+  const handleInputRefProp = useForkRef(inputRefProp, handleInputPropsRefProp);
+  const handleInputRef = useForkRef(inputRef, handleInputRefProp);
+
+  const [focused, setFocused] = React.useState(false);
+  const muiFormControl = React.useContext(FormControlContext);
+  const muiFormControlRef = React.useRef();
+  muiFormControlRef.current = muiFormControl;
```

Comment:
```suggestion

```

Reply:
It used to be done in an effect, what's the need for it? Can't we inline it?

Reply:
Having it in a ref means that it doesn't cause `checkDirty` to update and the `useEffect` to trigger and call the callbacks (`onEmpty` and `onFilled`). You could also do that by leaving it out of the useEffect dependencies and breaking the hook usage recommendations.

Reply:
checkDirty needs to access the latest values of the context, I believe.

Reply:
Ahh my bad, I was wrong on my other comment. As long as we are fine with the tests not being too representational we can inline it.